### PR TITLE
[SPARK-11100][SQL]HiveThriftServer HA issue,HiveThriftServer not registering with Zookeeper

### PR DESCRIPTION
**Pull request by xiaowangyu**
https://issues.apache.org/jira/browse/SPARK-11100

### CODE REVIEW

**Comment by WangTaoTheTonic** [comment]
with this patch, can we launch multiple thrift server instances and enable client to submit application without sensing particular thrift server addresses instead of zookeeper port?

**Comment by xiaowangyu** [comment]
Yes,jdbc url like this:
jdbc:hive2://zk1,zk2,zk3/default;serviceDiscoveryMode=zooKeeper;zooKeeperNamespace=sparkhiveserver2
see:https://issues.apache.org/jira/browse/HIVE-8376

**Comment by WangTaoTheTonic** [comment]
Yeah I think it is a nice feature which can bring HA into spark Thrift Server. Have you testet it yet?

**Comment by xiaowangyu** [comment]
Yes,My test is OK.

**Comment by xiaowangyu** [comment]
@liancheng can you help to review this PR? Thanks very much!

File: sql/hive-thriftserver/src/main/scala/org/apache/spark/sql/hive/thriftserver/HiveThriftServer2.scala [hunk 1]
```diff
@@ -57,6 +57,11 @@ object HiveThriftServer2 extends Logging {
     val server = new HiveThriftServer2(sqlContext)
     server.init(sqlContext.hiveconf)
     server.start()
+    if(SparkSQLEnv.hiveContext.hiveconf.getBoolVar(
```
Comment (andrewor14) [inline sql/hive-thriftserver/src/main/scala/org/apache/spark/sql/hive/thriftserver/HiveThriftServer2.scala]: need space after `if`, here and other places

**Comment by andrewor14** [comment]
Jenkins, this is ok to test.

@xiaowangyu can you describe your test setup briefly?
@liancheng if you have time please review.

**Comment by xiaowangyu** [comment]
Test this issue:
1.Add property to hive-site.xml
```
hive.server2.support.dynamic.service.discovery=true
hive.server2.zookeeper.namespace=sparkhiveserver2
hive.zookeeper.quorum=zk1,zk2,zk3
```
2.Start up sparkthriftserver
3.In the zookeeper path /sparkhiveserver2 will see:
```
serverUri=x.x.x.x:10000;version=1.2.1.spark;sequence=0000000026
```
4.Connect sparkthriftserver by beeline command:
```
!connect jdbc:hive2://zk1,zk2,zk3/default;serviceDiscoveryMode=zooKeeper;zooKeeperNamespace=sparkhiveserver2
```

**Comment by andrewor14** [comment]
ok to test

**Comment by searchwizard** [comment]
Will this be in Spark 1.6.1?

Thanks

**Comment by chutium** [comment]
+1 to this feature, may need to add this doc http://spark.apache.org/docs/latest/sql-programming-guide.html#running-the-thrift-jdbcodbc-server

**Comment by DoingDone9** [comment]
good job!

**Comment by xiaowangyu** [comment]
Any one can merge this patch?

**Comment by viper-kun** [comment]
@xiaowangyu 
1.Can you tell me how to work? Thrift server is easy to dump because of too many beelines. If beeline can choose less-load thrift server.

**Comment by xiaowangyu** [comment]
@viper-kun 
It is use zookeeper to choose the active thrift server when you connect by beeline or jdbc.
It only happen on the connect time. after connected, if the thrift server is down. the connection should be reconnect by client!

**Comment by viper-kun** [comment]
@xiaowangyu  What I means is how to choose which active thrift server.It chooses random or by thrift server load.

**Comment by xiaowangyu** [comment]
@viper-kun It's random!
You can check the code from hive:
https://github.com/apache/hive/blob/release-1.2.1/jdbc/src/java/org/apache/hive/jdbc/ZooKeeperHiveClientHelper.java#L53-L93

**Comment by rxin** [comment]
@xiaowangyu the thrift server has gone through a lot of changes lately, and we are not sure whether we'd want to keep the existing thrift server at all because it has a huge amount of technical debt. More so, I'm not sure if the zookeeper approach chosen by Hive is that great, since it is pretty difficult to operate in practice so I'm not sure if this would become the long-term solution. For now, I'd recommend you patching your own version of Spark to get this functionality.

**Comment by xiaowangyu** [comment]
@rxin 
Thanks! I get it.
BTW,any component will take the place of the thrift server?

**Comment by rxin** [comment]
We have currently inlined the Hive thrift server into the code base, but the long-term replacement is to be determined. I don't think anybody has thought a lot about it yet.

Do you mind closing this pull request for now? Thanks.

**Comment by viper-kun** [comment]
@rxin  Is there any design about replacement?

**Comment by rxin** [comment]
@viper-kun no - as I said, "I don't think anybody has thought a lot about it yet."

**Comment by xiaowangyu** [comment]
Thanks! I close it.

**Pull request closed without merging**